Updated `Gemspec/RequiredRubyVersion` to detect more issues

dvandersluis opened the pull request:
Adds detection for empty versions and `required_ruby_version` being set more than once. Also did a minor refactor to use `on_send` instead of having all the logic in `on_new_investigation`.

-----------------

Before submitting the PR make sure the following are checked:

* [x] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [x] Wrote [good commit messages][1].
* [ ] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Ran `bundle exec rake default`. It executes all tests and runs RuboCop on its own code.
* [x] Added an entry (file) to the [changelog folder](https://github.com/rubocop/rubocop/blob/master/changelog/) named `{change_type}_{change_description}.md` if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format) for details.

[1]: https://chris.beams.io/posts/git-commit/

koic commented on lib/rubocop/cop/gemspec/required_ruby_version.rb:
```diff
@@ -49,14 +57,15 @@ module Gemspec
-        def_node_search :required_ruby_version, <<~PATTERN
+        # @!method required_ruby_version?(node)
+        def_node_search :required_ruby_version?, <<~PATTERN
           (send _ :required_ruby_version= $_)
```
This capture looks unnecessary.

```suggestion
          (send _ :required_ruby_version= _)
```

dvandersluis replied:
Good call, it was left over from the previous version but I've removed it now 👍

koic commented on lib/rubocop/cop/gemspec/required_ruby_version.rb:
```diff
@@ -26,6 +28,12 @@ module Gemspec
+      #   Gem::Specification.new do |spec|
+      #     spec.required_ruby_version = '>= 2.5.0'
+      #     spec.required_ruby_version = '>= 2.5.0'
+      #   end
```
I think about it. It may be better to create a new cop like `Gemspec/DuplicateSpecificationAttribute` cop instead of this cop role extension. It seems preferable to keep one cop role simple based on Single Responsibility Principle :-) What do you think?

dvandersluis replied:
I think it makes sense all together because the single responsibility here is checking `required_ruby_version`, and it basically was already doing this, but just for 0 and 1 calls, so now extended to 2+ as well. 

However, if you feel strongly about this I would be fine splitting it (and I’d move the check for 0 calls to that new cop too).

koic replied:
IMHO, this cop is responsible for checking whether the values for `spec.required_ruby_version` of gemspec and `TargetRubyVersion` of .rubocop.yml match.

it seems simpler for another new cop to have the role of checking for duplicate attributes such as `required_ruby_version` for `Gem::Specification`. The new cop can check that various attributes such as `spec.name`, `spec.version`, `spec.description`, and etc are redundantly assigned.

bbatsov replied:
I agree with Koichi. Seems a separate cop looking for duplicated/empty properties would be a better idea.

dvandersluis replied:
So there actually is already another cop that checks for duplicate assignments: `Gemspec/DuplicatedAssignment` 🤦

dvandersluis commented:
I've updated this to just keep the added version checks (for `''` and `[]`, etc.) and removed the other logic. I've kept the refactoring to using `on_send` because I think that's preferable to doing everything in `on_new_investigation`
